pkg/typesystem/resolver_test.go

```diff
@@ -2,116 +2,196 @@ package typesystem
 
 import (
 	"context"
-	"fmt"
 	"testing"
 
 	"github.com/oklog/ulid/v2"
 	openfgav1 "github.com/openfga/api/proto/openfga/v1"
-	parser "github.com/openfga/language/pkg/go/transformer"
 	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
 	"go.uber.org/mock/gomock"
-	"golang.org/x/sync/errgroup"
 
 	mockstorage "github.com/openfga/openfga/internal/mocks"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/testutils"
 )
 
 func TestMemoizedTypesystemResolverFunc(t *testing.T) {
-	mockController := gomock.NewController(t)
-	defer mockController.Finish()
-
-	mockDatastore := mockstorage.NewMockOpenFGADatastore(mockController)
-
-	storeID := ulid.Make().String()
-	modelID1 := ulid.Make().String()
-	modelID2 := ulid.Make().String()
-
-	typedefs := parser.MustTransformDSLToProto(`
-		model
-			schema 1.1
-		type user
-		type document
-			relations
-				define viewer: [user]`).GetTypeDefinitions()
-
-	gomock.InOrder(
-		mockDatastore.EXPECT().
-			ReadAuthorizationModel(gomock.Any(), storeID, modelID1).
-			Return(&openfgav1.AuthorizationModel{
-				Id:              modelID1,
-				SchemaVersion:   SchemaVersion1_1,
-				TypeDefinitions: typedefs,
-			}, nil),
-
-		mockDatastore.EXPECT().
-			FindLatestAuthorizationModel(gomock.Any(), storeID).
-			Return(&openfgav1.AuthorizationModel{
-				Id:              modelID2,
-				SchemaVersion:   SchemaVersion1_1,
-				TypeDefinitions: typedefs,
-			}, nil),
-	)
-
-	resolver, resolverStop := MemoizedTypesystemResolverFunc(
-		mockDatastore,
-	)
-	defer resolverStop()
-
-	typesys, err := resolver(context.Background(), storeID, modelID1)
-	require.NoError(t, err)
-	require.NotNil(t, typesys)
-
-	relation, err := typesys.GetRelation("document", "viewer")
-	require.NoError(t, err)
-	require.NotNil(t, relation)
-
-	typesys, err = resolver(context.Background(), storeID, "")
-	require.NoError(t, err)
-	require.NotNil(t, typesys)
-	require.Equal(t, modelID2, typesys.GetAuthorizationModelID())
-
-	relation, err = typesys.GetRelation("document", "viewer")
-	require.NoError(t, err)
-	require.NotNil(t, relation)
-}
-
-func TestSingleFlightMemoizedTypesystemResolverFunc(t *testing.T) {
```

Comment:
Note: I deleted this test because it doesn't prove that singleflight works. I removed the singleflight code and this test still passes. This is because line 93 `.MinTimes(1).MaxTimes(numGoroutines)` works with and without singleflight.